### ISSUE TITLE
[device-port] [mata] Port to Essential Ph-1

### ISSUE THREAD

**Issue by xonq** [Halium/projectmanagement#122]
Tree: 

- [X] Create manifest 
- [x] Boot image and system image build successfully
- [ ] Device boots into rootfs, `usb: Manufacturer: GNU/Linux Device` appears in `dmesg` on host.
- [ ] LXC container starts and does not crash
- [ ] libhybris tests
  * [ ] test_gps
  * [ ] test_hwcomposer
  * [ ] test_lights
  * [ ] test_vibrator
  * [ ] test_wifi
  * [ ] test_sensors
  * [ ] test_audio
  * [ ] test_camera
  * [ ] test_input
  * [ ] test_nfc
  * [ ] test_recorder

**Comment by xonq** [comment]
Following enabling envsetup.sh, I get this error 'Can not locate config makefile for product "lineage_mata". It exists in BUILDDIR/halium/device/essential/mata.

`breakfast mata
including vendor/cm/vendorsetup.sh
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.
build/core/product_config.mk:249: *** Can not locate config makefile for product "cm_mata".  Stop.
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.
Device mata not found. Attempting to retrieve device repository from LineageOS Github (http://github.com/LineageOS).
Found repository: android_device_essential_mata
Traceback (most recent call last):
  File "vendor/cm/build/tools/roomservice.py", line 263, in <module>
    default_revision = get_default_revision()
  File "vendor/cm/build/tools/roomservice.py", line 117, in get_default_revision
    return r.replace('refs/heads/', '').replace('refs/tags/', '')
AttributeError: 'NoneType' object has no attribute 'replace'
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.

** Don't have a product spec for: 'lineage_mata'
** Do you have the right repo manifest?`

**Comment by xonq** [comment]
Attempting to setup remote for unofficial lineage 16.0, Invisiblek's, in the manifest.

**Comment by xonq** [comment]
'breakfast' command still failing - it isnt finding my lineage_mata.mk file even though it exists

**Comment by jbruechert** [comment]
Can you post a full log? Maybe some dependency makefile is missing.

**Comment by xonq** [comment]
```
including vendor/cm/vendorsetup.sh
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.
build/core/product_config.mk:249: *** Can not locate config makefile for product "cm_mata".  Stop.
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.
Device mata not found. Attempting to retrieve device repository from LineageOS Github (http://github.com/LineageOS).
Found repository: android_device_essential_mata
Default revision: lineage-16.0
Checking branch info
Checking if device/essential/mata is fetched from android_device_essential_mata
LineageOS/android_device_essential_mata already fetched to device/essential/mata
Syncing repository to retrieve project.
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0        
fatal: remote github not defined in /home/xaq/halium/.repo/manifest.xml
Repository synced!
Looking for dependencies in device/essential/mata
Done
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.
build/core/product_config.mk:249: *** Can not locate config makefile for product "lineage_mata".  Stop.

** Don't have a product spec for: 'lineage_mata'
** Do you have the right repo manifest?
```

I put the dependencies from lineagedependencies into the manifest. cm_mata doesn't exist, but lineage_mata does in halium/device/essential/mata

**Comment by xonq** [comment]
```
~/halium$ breakfast
including vendor/cm/vendorsetup.sh

You're building on Linux

Breakfast menu... pick a combo:
 1. full-eng 
... and don't forget the bacon!

Which would you like? [aosp_arm-eng] 

```

Here's my breakfast command output as well

**Comment by jbruechert** [comment]
Are you trying to use a lineage-16.0 device tree? The halium tree is currently 14.1 based, so this will not work!

**Comment by xonq** [comment]
Worked

**Comment by xonq** [comment]
Curious about how functional it'll be on cm 14.1. 16.0 works fine but 14.1 hasn't been edited in a year

**Comment by ghost** [comment]
I'm not sure if you've had any luck, but I was able to get both the hybris-boot and system images built, but have run into some issues getting them installed because of the slots

**Comment by alefnode** [comment]
I found a problem when make mka hybris-boot

> [ 32% 675/2049] Building Kernel Config
> FAILED: /bin/bash -c "(make -j8  CFLAGS_MODULE=\"-fno-pic\" -C kernel/essential/msm8998 O=/home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ ARCH=arm64 CROSS_COMPILE=\" aarch64-linux-androidkernel-\" VARIANT_DEFCONFIG= SELINUX_DEFCONFIG= lineageos_mata_defconfig ) && (if [ ! -z \"\" ]; then                        echo \"Overriding kernel config with ''\";                      echo  >> /home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ/.config;                       make -j8 -C kernel/essential/msm8998 O=/home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ ARCH=arm64 CROSS_COMPILE=\" aarch64-linux-androidkernel-\" oldconfig; fi ) && (make -j8 -C kernel/essential/msm8998 O=/home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ ARCH=arm64 CROSS_COMPILE=\" aarch64-linux-androidkernel-\" savedefconfig ) && (if [ ! -z \"\" ]; then                      echo \"Using additional config ''\";                    kernel/essential/msm8998/scripts/kconfig/merge_config.sh -m -O /home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ /home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ/.config kernel/essential/msm8998/arch/arm64/configs/;                         make -j8 -C kernel/essential/msm8998 O=/home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ ARCH=arm64 CROSS_COMPILE=\" aarch64-linux-androidkernel-\" KCONFIG_ALLCONFIG=/home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ/.config alldefconfig; fi )"
> make: se entra en el directorio '/home/debian/halium/kernel/essential/msm8998'
> make[1]: se entra en el directorio '/home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ'
>   HOSTCC  scripts/basic/fixdep
>   GEN     ./Makefile
>   HOSTCC  scripts/kconfig/conf.o
>   SHIPPED scripts/kconfig/zconf.tab.c
>   SHIPPED scripts/kconfig/zconf.lex.c
>   SHIPPED scripts/kconfig/zconf.hash.c
>   HOSTCC  scripts/kconfig/zconf.tab.o
>   HOSTLD  scripts/kconfig/conf
> drivers/staging/Kconfig:113: can't open file "drivers/staging/qcacld-3.0/Kconfig"
> /home/debian/halium/kernel/essential/msm8998/scripts/kconfig/Makefile:110: fallo en las instrucciones para el objetivo 'lineageos_mata_defconfig'
> make[2]: *** [lineageos_mata_defconfig] Error 1
> /home/debian/halium/kernel/essential/msm8998/Makefile:535: fallo en las instrucciones para el objetivo 'lineageos_mata_defconfig'
> make[1]: *** [lineageos_mata_defconfig] Error 2
> make[1]: se sale del directorio '/home/debian/halium/out/target/product/mata/obj/KERNEL_OBJ'
> Makefile:146: fallo en las instrucciones para el objetivo 'sub-make'
> make: *** [sub-make] Error 2
> make: se sale del directorio '/home/debian/halium/kernel/essential/msm8998'
> [ 32% 675/2049] target asm: libc_syscalls <= bionic/libc/arch-arm64/syscalls/_exit.S
> ninja: build stopped: subcommand failed.
> build/core/ninja.mk:151: fallo en las instrucciones para el objetivo 'ninja_wrapper'
> make: *** [ninja_wrapper] Error 1
> make: se sale del directorio '/home/debian/halium'


Anyone else with this error?

**Comment by alefnode** [comment]
Some tips:
sudo apt install libssl-dev

Comment line 113 from ./kernel/essential/msm8998/drivers/staging/Kconfig

> #source "drivers/staging/qcacld-3.0/Kconfig"

**Comment by ghost** [comment]
Yeah there are definitely a couple weird little things about the build. With those two tips, were you able to finish the hybris-boot build?

**Comment by alefnode** [comment]
Yeah "mka hybris-boot" works with that tips, now i'm stacked on "mka systemimage" with this error:

> Running kati to generate build-lineage_mata.ninja...
> No need to regenerate ninja file
> Starting build with ninja
> ninja: Entering directory `.'
> ninja: error: '/home/debian/halium/out/host/linux-x86/bin/build_verity_tree', needed by '/home/debian/halium/out/target/product/mata/obj/PACKAGING/systemimage_intermediates/system.img', missing and no known rule to make it
> build/core/ninja.mk:151: fallo en las instrucciones para el objetivo 'ninja_wrapper'
> make: *** [ninja_wrapper] Error 1
> make: se sale del directorio '/home/debian/halium'
>

**Comment by alefnode** [comment]
Modify file: build/target/product/verity.mk
Set PRODUCT_SUPPORTS_VERITY := false
And delete verity_key from PRODUCT_PACKAGES

**Comment by alefnode** [comment]
It works! you can use my last tips to compile halium on essential phone

**Comment by alefnode** [comment]
Ok, after flashing hybris-boot and system.img i've got a bootloop, y tried to debug it but i can't

**Comment by alefnode** [comment]
https://paste.ubuntu.com/p/VSH3dKhgzd/

**Comment by Maxr1998** [comment]
The reboot/bootloop seems to be because of dm-verity, did you try to disable that? Afaik you have to remove the `verify` mount option inside the fstab.

**Comment by xonq** [comment]
keep getting this when running mka halium-boot for each of the dependencies in fixdep.c

`fatal error: unistd.h: No such file or directory`

**Comment by xonq** [comment]
fixed with

'sudo apt install libc6-dev'

**Comment by xonq** [comment]
New error on system image. Got hybris-boot fine.

`[  0% 7/20168] Building Kernel Config
make: Entering directory '/home/xaq/Desktop/halium/kernel/essential/msm8998'
make[1]: Entering directory '/home/xaq/Desktop/halium/out/target/product/mata/obj/KERNEL_OBJ'
  GEN     ./Makefile
warning: (SND_SOC_WCD934X) selects SND_SOC_WCD_SPI which has unmet direct dependencies (SOUND && !M68K && !UML && SND && SND_SOC && CONFIG_SPI)
warning: (SND_SOC_WCD934X) selects SND_SOC_WCD_SPI which has unmet direct dependencies (SOUND && !M68K && !UML && SND && SND_SOC && CONFIG_SPI)
#
# configuration written to .config
#
make[1]: Leaving directory '/home/xaq/Desktop/halium/out/target/product/mata/obj/KERNEL_OBJ'
make: Leaving directory '/home/xaq/Desktop/halium/kernel/essential/msm8998'
make: Entering directory '/home/xaq/Desktop/halium/kernel/essential/msm8998'
make[1]: Entering directory '/home/xaq/Desktop/halium/out/target/product/mata/obj/KERNEL_OBJ'
  GEN     ./Makefile
scripts/kconfig/conf  --savedefconfig=defconfig Kconfig
warning: (SND_SOC_WCD934X) selects SND_SOC_WCD_SPI which has unmet direct dependencies (SOUND && !M68K && !UML && SND && SND_SOC && CONFIG_SPI)
warning: (SND_SOC_WCD934X) selects SND_SOC_WCD_SPI which has unmet direct dependencies (SOUND && !M68K && !UML && SND && SND_SOC && CONFIG_SPI)
make[1]: Leaving directory '/home/xaq/Desktop/halium/out/target/product/mata/obj/KERNEL_OBJ'
make: Leaving directory '/home/xaq/Desktop/halium/kernel/essential/msm8998'
ninja: build stopped: subcommand failed.
build/core/ninja.mk:151: recipe for target 'ninja_wrapper' failed
make: *** [ninja_wrapper] Error 1
make: Leaving directory '/home/xaq/Desktop/halium'
`

**Comment by xonq** [comment]
Will this manifest work - it's based on the 16.0 dependencies. If you look at invisiblek's cm-14.1 lineageos repo, the dependencies are much more extensive. I'm going to bake that into my manifest and work from that.

**Comment by xonq** [comment]
my systemimage keeps failing with

`warning: (SND_SOC_WCD934X) selects SND_SOC_WCD_SPI which has unmet direct dependencies (SOUND && !M68K && !UML && SND && SND_SOC && CONFIG_SPI)
warning: (SND_SOC_WCD934X) selects SND_SOC_WCD_SPI which has unmet direct dependencies (SOUND && !M68K && !UML && SND && SND_SOC && CONFIG_SPI)`

**Comment by xonq** [comment]
fixed and successfully built systemimage by running `make clean` after building the boot img and used `export USE_HOST_LEX=yes`

disabled 'verify' in the fstab and set CONFIG_DEFAULT_SECURITY to apparmor in lineageos_mata_defconfig, also adjusted CONFIG_SECURITY_xx_BOOTPARAM to 1 for SELINUX and 0 for apparmor

will flash soon and report back

**Comment by xonq** [comment]
Edit 190712: I believe to be caused by a mistake building, so this may be ignorable

kernel panic caused by SE Linux: https://paste.ubuntu.com/p/dGDsPmxjPv/
(please note I'm working to boot ubuntu touch so am modding files with their tools)

here are my defconfig se linux parameter values:
`CONFIG_SECURITY_SELINUX_BOOTPARAM_VALUE=1
CONFIG_SECURITY_SELINUX=y
CONFIG_SECURITY_SELINUX_BOOTPARAM=y
CONFIG_SECURITY_SELINUX_DISABLE=y
CONFIG_SECURITY_SELINUX_DEVELOP=y
CONFIG_SECURITY_SELINUX_AVC_STATS=y
CONFIG_SECURITY_SELINUX_CHECKREQPROT_VALUE=1`

I rebuilt halium-boot after changing the bootparam, selinux, develop, and stats to `n`. Caused a watchdog bite after flashing, otherwise output was the same.

**Comment by alefnode** [comment]
Did you see this?

[    1.696244] init: No precompiled sepolicy: No such file or directory
[    1.696255] init: Compiling SELinux policy
[    1.696386] init: Failed to read /vendor/etc/selinux/plat_sepolicy_vers.txt: No such file or director

**Comment by xonq** [comment]
That's exactly what I saw before the panic

ub ports does not use se linux though, correct? but should I set the default security parameter from apparmor back to selinux?

**Comment by alefnode** [comment]
Are you finally working with cm14.1?

Can you share your manifest?

**Comment by alefnode** [comment]
How did you disable verify?, i can't pass over that point

**Comment by xonq** [comment]
I'm using the same manifest as provided at this point. I will begin working with invisiblek's cm-14.1 dependencies if these dont follow through. I have disabled verify, will flash tonight to check - ran into some issues flashing. 

Got to have a USB 2 to C cable and port, the right fastboot, and all that

**Comment by xonq** [comment]
I disabled verify per the advice above. Edit: `/halium/device/essential/mata/rootdir/root/fstab.mata` From there you will see a parameter on the first line of sources that says `verify`, simply delete that

**Comment by xonq** [comment]
I disabled verify and got the same error you were receiving alefnode about dm-verity failing.
Went into the defconfig and commented out 3 lines dealing with verity.
The new error is something to do with dm-0 Edit: `see alefnode's paste 7 comments down for details`. 
EDIT: built from the old dependencies in lineagos-cm-14.1 (largely @invisiblek's builds and hardware trees successfully below

**Comment by alefnode** [comment]
I made some changes:
**File ./device/essential/mata/device.mk comment lines**
#PRODUCT_SYSTEM_VERITY_PARTITION := /dev/block/bootdevice/by-name/system
#$(call inherit-product, build/target/product/verity.mk)
**File ./kernel/essential/msm8998/arch/arm64/configs/mata_defconfig** 
CONFIG_DM_VERITY=n
CONFIG_DM_VERITY_FEC=n
CONFIG_DM_ANDROID_VERITY=n

But still not working

**Comment by xonq** [comment]
> I made some changes:
> **File ./device/essential/mata/device.mk comment lines**
> #PRODUCT_SYSTEM_VERITY_PARTITION := /dev/block/bootdevice/by-name/system
> #$(call inherit-product, build/target/product/verity.mk)
> **File ./kernel/essential/msm8998/arch/arm64/configs/mata_defconfig**
> CONFIG_DM_VERITY=n
> CONFIG_DM_VERITY_FEC=n
> CONFIG_DM_ANDROID_VERITY=n
> 
> But still not working

have you tried messing with the lineage mata defconfig?
edit: /halium/kernel/essential/msm8998/arch/arm64/configs/lineageos_mata_defconfig

**Comment by alefnode** [comment]
No, i am going to did it now, for checking

**Comment by alefnode** [comment]
I tried all, all the change commented on this post, but i continue with same error

Fstab + device.mk + verity.mk + defconfig

https://paste.ubuntu.com/p/7FTjQJ6FzF/

**Comment by xonq** [comment]
did you verify that the variables were actually changed in the final .config file? Should be located in the /out folder I'm guessing

**Comment by alefnode** [comment]
```
Yeah, i have this

fs_type=ext4
system_size=4294967296 
extfs_sparse_flag=-s
squashfs_sparse_flag=-s
selinux_fc=/home/debian/halium/out/target/product/mata/root/file_contexts.bin
recovery_as_boot=true
system_root_image=true
ramdisk_dir=/home/debian/halium/out/target/product/mata/root
skip_fsck=true
```
Ignore this comment, i think i was using wrong defconfig file, testing again.

**Comment by alefnode** [comment]
Ok, finally i pass to next step:

https://paste.ubuntu.com/p/fV597YgDGx/

**Comment by xonq** [comment]
I commented out the verity options in `device.mk` in the /halium/build directory and rebuilt. Will update later

**Comment by xonq** [comment]
That didn't work. Perhaps we should be focusing on making verity work instead of disabling it.

**Comment by xonq** [comment]
I'm failing at building `build_verity_tree` for systemimage. I'm not sure how to build that binary. `mka build_verity_tree` doesn't work

**Comment by Maxr1998** [comment]
It seems you have to build that one "manually" with `cryptsetup`, as it's described here: https://source.android.com/security/verifiedboot/dm-verity#hash-tree

I love what you guys are doing here, and hope you'll get it to work eventually. I'd contribute if my PH-1 wasn't my daily driver..

**Comment by xonq** [comment]
Perfect, will follow up with that.
I started using the `lineage-dependencies` from the cm-14.1 tree for my repo manifest because I noticed quite a few hardware connections were failing in the boot notes. @Invisiblek has specific hardware dependencies in his repository in the cm-14.1 tree.
Edit: The only difference in building is that the `opensource-nfc..` (near field communication) module fails to build due to missing a java_library, so I omitted that at this point. I assume this manifest will be better because it specifies more dependencies directly... the qcacld3.0 built, which was an error in the previous manifest.

Edit: cm 14.1 manifest:
https://pastebin.com/2zqCT2VM

**Comment by xonq** [comment]
I tried following those instructions to build and sign `verity`, but it wasn't working for the boot.img successfully. At this time, I found a simple [patch](https://gist.github.com/Sergobot/e201cfcbc13eae3c3c9fef6ade9ad185) for the verity option I'm going to try.

**Comment by alefnode** [comment]
Awesome manifest, i would try that too.

The patch is same thing that we mention before, so i think it is not going to work

**Comment by xonq** [comment]
You're right, it didnt. So, as I see it:

- try to bypass verity
  - magisk bypasses some verity, maybe we could try that too?
- try to sign it correctly

FYI if you're having trouble with twrp touch functionality you have to:

Reboot fastboot
run `flashall` from factory image download
Flash `twrp.img` to ONE slot 
Change to slot you just flashed, reboot recovery
run `adb shell twrp install twrp-installer.zip`, Reboot fastboot
Change to other slot, twrp is functional on this slot

**Comment by alefnode** [comment]
Yeah, i have touch working at first i was crazy about that.

Do you think this can help us? 
https://forum.xda-developers.com/android/software-hacking/signing-boot-images-android-verified-t3600606

**Comment by xonq** [comment]
I tried that and it didnt work, however I changed the key name to verity. Might be worth you trying!

**Comment by ghost** [comment]
So I ended up cheating a bit and just extracted the tarball version of this repo into the out/host/linux-x86/bin folder and renamed `build_verity_tree.py` to `build_verity_tree` https://android.googlesource.com/platform/system/extras/+/724396ed8dbf71ce2744e916fec2e6aec05d1041/verity/

Good news is that it worked well enough to not fail there again. 
Bad news is it's then failing and calling for a new binary file `fec`.

I haven't been able to find that yet, but if I can, I'll see what happens with the same method. 


Potentially also of note, I found Essential's Github page: https://github.com/EssentialOpenSource where they have a repo for qcacld-3.0

**Comment by xonq** [comment]
@laserbeamh great news, that's a significant hurdle overcome. 

I've reverted back to Essential being my daily driver, but I can do some research in the meantime before I move forward with development again.

**Comment by xonq** [comment]
Perhaps [this](https://forum.xda-developers.com/android/software/universal-dm-verity-forceencrypt-t3817389) will work. 

- Flash stock in bootloader (download from essential, run 'flashall' or do manually)
- Wipe all `fastboot -w`
- Flash halium-boot.img to 'a' `fastboot flash boot_a halium-boot.img`
- Flash twrp to 'b' `fastboot flash boot_b twrp.img`
- Boot recovery on 'b' `fastboot -ab` boot recovery
- Install twrp zip `adb shell twrp install twrp.zip`
- Reboot recovery 'a' `adb reboot bootloader` `fastboot -aa` boot recovery
- Flash the disable dm-verity and fec zip `adb shell twrp install disable.zip`

**Comment by ghost** [comment]
took the lazy route (again). eventually i'll go back and at least set git dependencies correctly.

this is the repo for fec: https://android.googlesource.com/platform/system/extras/+/master/verity/fec/tests/

`mka systemimage` failed at 7%

`[  7% 1572/22028] Lex: libsepol_32 <= external/selinux/libsepol/cil/src/cil_lexer.l
FAILED: /bin/bash -c "prebuilts/misc/linux-x86/flex/flex-2.5.39 -o/home/laserbeamh/halium/out/host/linux-x86/obj32/SHARED_LIBRARIES/libsepol_intermediates/cil/src/cil_lexer.c external/selinux/libsepol/cil/src/cil_lexer.l"
flex-2.5.39: loadlocale.c:130: _nl_intern_locale_data: Assertion cnt < (sizeof (_nl_value_type_LC_TIME) / sizeof (_nl_value_type_LC_TIME[0]))' failed.
Aborted (core dumped)
[  7% 1572/22028] host C: libsqlite_32 <= external/sqlite/dist/sqlite3.c
ninja: build stopped: subcommand failed.
build/core/ninja.mk:151: recipe for target 'ninja_wrapper' failed
make: *** [ninja_wrapper] Error 1
`

**Comment by ghost** [comment]
yeah that one's on me. used the flex fix from the docs and made it passed that. had to install `libxml2-utils` for a new issue. now stuck on an issue with verity sizing.

https://gist.github.com/laserbeamh/e178450ae67e9bacac054807581cadce

**Comment by xonq** [comment]
Not sure I'm following what you're doing here. You successfully built and bypassed the verifying dm-verity step and now you're getting hung up on rebuilding if I understand right. What did you change?

**Comment by ghost** [comment]
Sorry, I've been a little all over the place... I'm also still trying to find a good way to document for others instead of just my scrambled brain. In a nutshell:

- I got to where I think your build was failing with the bit about there not being the `build_verity_tree` bin file and no know rules to make it. 
- I saw what you guys had been discussing and ended up finding `build_verity_tree.py` on Google's git page so I took that file and put it in my `bin` folder, tried recompiling and ran into a similar issue with missing an `fec` bin.
- Found that file in a different repo on Google's git in a folder named `verity` (because of course)
- Had to fix my `flex` locale and install `libxml2-utils` both of which made make attempts fail
- got all of that squared away and now build makes it to 99% (from a cold build) and fails 

[gist of my build fail now](https://gist.github.com/laserbeamh/775a5d865ac7e1f14af782bf45201fab). I'm seeing two different possible issues and can't tell which is causing the fail. There's this bit on the fs type and size: `build_verity_metadata.py: error: invalid choice: '4294967296' (choose from 'size', 'build')                                                                                                                                                                                    
fs type is not ext4` 

and this bit with it being an issue with sizing `Could not build verity tree! Error:                                                                                                                                                                                                                                            
Error: making verity enabled image failed                                                                                                                                                                                                                                      
Out of space? the tree size of /home/avery/halium/out/target/product/mata/system is (MB):                                                                                                                                                                                      
379     /home/avery/halium/out/target/product/mata/system                                                                                                                                                                                                                      
The max is 4096 MB`. 

I'm not really sure where to go from here, but I think once this is sorted out, the build may succeed with a verity signature! Thoughts?

**Comment by xonq** [comment]
Verity seems ideal - for the sake of getting a working development model going, I'm fine without it, which is why I've suggested the disable.zip above that supposedly disables verity and fec.

Maybe your verity build may be failing because the file is drawn from a different device and expecting a different size? - why wouldn't that happen the first time though?

**Comment by ghost** [comment]
My knowledge of all of this is shaky. I'm not against disabling verity and fec, but I know it would nag at me and it feels so close :/

Right now I'm going to see if I can find a way to convince verity that the fs is ext4. In the makefile there's an if statement where if the user images use ext4 then the max size is the partition size (4gb) but the else statement is the 4gb/1048576 leading to the 4096mb. The only other spot with what I grepped for was in a ninja file.

My thinking/current mission is if I can find where that issue is, then I can actually get the build. If I can't get it in the near future, I'll try the build with the zip you suggested and see if that builds for now.

Also going to try to add back in the qcald from essential to see if I can get that going correctly too!

**Comment by xonq** [comment]
Thats great. I look forward to when you report back

**Comment by jrobinson52** [comment]
Any updates on this? I want to help

**Comment by xonq** [comment]
> Any updates on this? I want to help

No updates on my end - it's my daily driver still.

**Comment by alefnode** [comment]
> > Any updates on this? I want to help
> 
> No updates on my end - it's my daily driver still.

Same here

**Comment by jrobinson52** [comment]
I've restarted my work on this but I've run into a problem. I've gone to flash the built images to my phone but after flashing hybris-boot.img I can't get into recovery to run the halium-install script against it. Am I missing a step somewhere in the documentation?

**Comment by doniks** [comment]
That is surprising. Flashing Boot should not influence the workings of recovery. But then, I see talk of A/B above, and I've never touched such a device.  I guess you have seen this:  https://github.com/Halium/projectmanagement/issues/122#issuecomment-525831091
Does that help at all? How exactly did you flash boot? How are you trying to enter recovery? Can you fastboot boot twrp.img ?

**Comment by jrobinson52** [comment]
I re-flashed the stock image before starting this process. I follow the steps in #122 but when I go to boot into recovery (via fastboot command on phone or by rebooting and holding PWR & VOL+) the phone either reboots into fastboot mode or has the Essential logo and some small text at the top corner that says 'Press a button to shutdown'. I tried doing `fastboot boot twrp.img` but I get an error: `FAILED (remote: unknown command)` Although now I can't boot into TWRP at all following the TWRP documentation either.

**Comment by doniks** [comment]
> I tried doing `fastboot boot twrp.img` but I get an error: `FAILED (remote: unknown command)` 

But you are sure that you ARE in fastboot mode right? When you say `fastboot devices` you get a proper response?

**Comment by jrobinson52** [comment]
Yes, I am in fastboot mode and it shows up normally when running `fastboot devices`

**Comment by doniks** [comment]
mhm ... bootloader is unlocked?

**Comment by jrobinson52** [comment]
Yes, unless flashing stock re-locks it.

**Comment by jrobinson52** [comment]
Update: TWRP boots after flashing but when it does it simply displays the error in the top left saying to press a key to shutdown

**Comment by jrobinson52** [comment]
I got ahold of another phone, I'm able to flash twrp normally to it. Looks like the phone I was using gave up the ghost. Unfortunately, after flashing twrp to slot a, flashing halium-boot to slot b, switching back to slot a to run halium install, then switching back to slot b to boot the installation, I simply get a boot loop. 

There's no information in the dmesg output and the screen on this phone is cracked so if there's any pertinent info on screen in the corner I can't see it. Any ideas?

**Comment by dragynbane222** [comment]
If it helps, starting tomorrow my Mata won't be my DD, and it's in good condition, so I can put mine towards texting. 

> I got ahold of another phone, I'm able to flash twrp normally to it. Looks like the phone I was using gave up the ghost. Unfortunately, after flashing twrp to slot a, flashing halium-boot to slot b, switching back to slot a to run halium install, then switching back to slot b to boot the installation, I simply get a boot loop.
> 
> There's no information in the dmesg output and the screen on this phone is cracked so if there's any pertinent info on screen in the corner I can't see it. Any ideas?

If it helps, starting tomorrow, my. Mata won't be my DD, so I can put it towards testing. It's in good condition (only a small crack on the back on one of the corners), so I should be able to get a shot of what shows up on the screen

**Comment by jrobinson52** [comment]
Thanks for the help. If you want my current build, I have links for the [halium-boot.img](https://nextcloud.jrrobinson.site/index.php/s/rsF9MXHj2gtY687) and the [system.img](https://nextcloud.jrrobinson.site/index.php/s/aZKbZJ4Ttnf9K7a)

(Both links expire this Saturday May 9)

**Comment by dragynbane222** [comment]
> Thanks for the help. If you want my current build, I have links for the [halium-boot.img](https://nextcloud.jrrobinson.site/index.php/s/rsF9MXHj2gtY687) and the [system.img](https://nextcloud.jrrobinson.site/index.php/s/aZKbZJ4Ttnf9K7a)
> 
> (Both links expire this Saturday May 9)

So I just tested, there's no meaningful output, simply the Essential boot logo on every attempt at boot. Sorry I couldn't find anything more meaningful

**Comment by jrobinson52** [comment]
Thanks anyways. I'd love to keep working on this but now my working phone is exhibiting the same issues as the non-working one. I've tried both Windows and linux to see if that was the issue along with multiple cables. Not being able to flash twrp puts me at a stand-still for now. I'll get back to you if I have any progress

**Comment by doniks** [comment]
> Yes, unless flashing stock re-locks it.

does it?

**Comment by jrobinson52** [comment]
I finally have my phone working again. Turns out there's some oddities with an android 10 installation and TWRP. I've also learned the distinct difference between `fastboot continue` and `fastboot reboot`. With that sorted out, I'm back to working on it and will keep you posted on any updates

**Comment by Veisen** [comment]
every version of halium 7,8 and 9 do not boot on essential ...... i have problem with security every time

same problem as on pixel phones

**Comment by alfonsojon** [comment]
I am currently attempting a build using some cm-14.1 trees from InvisibleK but the rest from Lineage OS. I ran into an issue where it could not locate `java_library.mk`, so I symbolically linked it to `static_java_library.mk`. I was then able to build the boot image as well as hybris image. Currently working on pulling the proper proprietary files so building the system image is happy.

**Comment by alfonsojon** [comment]
https://gist.github.com/alfonsojon/f8162c5f7b35346f23ae3711ff79e91f 
Here's the list of files that the shell script could not locate on the factory Android 7.1.1 image. After using `adb shell` then `su`, it is able to find the files where they should be. Seems like I just need to figure out why it's not pulling files.

Edit: what is easiest is mounting the essential factory image after decompressing it using `simg2img`, then feeding that script to `extract-files.sh`

Edit 2: [remaining missing files](https://gist.github.com/alfonsojon/5a3b4c80d38261bf6fabf1ca251ae7b7)

**Comment by alfonsojon** [comment]
I disabled verity compatibility in build/target/product/verity.mk, and am using TheMuppets vendor blobs as my phone seems to be missing some of them. Building system image now.

**Comment by alfonsojon** [comment]
I succeeded in building system.img; however, I accidentally bricked my phone when flashing stock 7.1 using a low-quality USB cable. D'oh!

**Comment by Veisen** [comment]
can you upload your images i can try it

**Comment by alfonsojon** [comment]
> can you upload your images i can try it

[Link to hybris-boot.img](https://mega.nz/file/0s9gAQ7a#WIyCYoDx5LKjxWgg7jVi9t04TxS6LXtbbCbXd6lQFgs)
`sha256sum: ba620b2aa819dc1ada62bcb9c40994a32b8fbaf5167e2203975c4f45683c0fac`
[Link to system.img](https://mega.nz/file/ZhlixaQR#3Hnwbn_T2kBqfsj4UHpgj0rfhZvnRXvzxS0BxjY2R40)
`sha256sum: 0caf5147796d82c48d1b71f00c81e2d9a37c66ea0f123ee95c7bf4d95f462e37`

you still have to get the rootfs but those are the images needed to run it. Note, I have not tested them yet as my PH-1 is bricked.

**Comment by Veisen** [comment]
what is frimware base of your build ?

**Comment by Veisen** [comment]
> > can you upload your images i can try it
> 
> [Link to hybris-boot.img](https://mega.nz/file/0s9gAQ7a#WIyCYoDx5LKjxWgg7jVi9t04TxS6LXtbbCbXd6lQFgs)
> `sha256sum: ba620b2aa819dc1ada62bcb9c40994a32b8fbaf5167e2203975c4f45683c0fac`
> [Link to system.img](https://mega.nz/file/ZhlixaQR#3Hnwbn_T2kBqfsj4UHpgj0rfhZvnRXvzxS0BxjY2R40)
> `sha256sum: 0caf5147796d82c48d1b71f00c81e2d9a37c66ea0f123ee95c7bf4d95f462e37`
> 
> you still have to get the rootfs but those are the images needed to run it. Note, I have not tested them yet as my PH-1 is bricked.

bootloops like on my builds

**Comment by alfonsojon** [comment]
> what is frimware base of your build ?

CM-14.1
